### PR TITLE
Plugin/poopmap

### PR DESCRIPTION
**Pull request by matievisthekat**
### PoopMap plugin

This plugin displays a user's pooping statistics from [PoopMap](https://poopmap.net)

An example can be viewed at [here](https://github.com/marketplace/actions/poopmap-stats)

### CODE REVIEW

**Comment by matievisthekat** [comment]
Hi, I'm having trouble testing the plugin. Is there some sort of guide? Otherwise, how do I enable the plugin for use on the web instance?

**Comment by Nixinova** [comment]
> PoopMap

🤨

**Comment by matievisthekat** [comment]
PoopMap: https://poopmap.net

File: source/plugins/poopmap/index.mjs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      return null
+
+    const {token, days} = imports.metadata.plugins.poopmap.inputs({data, account, q})
+    const {data:{poops}} = await imports.axios.get(`https://api.poopmap.net/api/v1/public_links/${token}`)
+
+    const filteredPoops = poops.filter(poop => {
+      const createdAt = new Date(poop.created_at)
+      const now = new Date().getTime()
+      //Days * hours * minutes * seconds * milliseconds
+      const timeframe = now - (days ?? 7) * 24 * 60 * 60 * 1000
```
Comment (lowlighter) [inline source/plugins/poopmap/index.mjs:16]:
```suggestion
      const timeframe = now - days * 24 * 60 * 60 * 1000
```
`days` will already defaults to `7` as it goes through the input validator

File: source/plugins/poopmap/index.mjs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+      poop.created_at = createdAt.toString()
+
+      return createdAt.getTime() > timeframe
```
Comment (lowlighter) [inline source/plugins/poopmap/index.mjs:20]: I'm pretty sure you can directly do `new Date(poop.created_at) > new Date(Date.now() - days * 24 * 60 * 60 * 1000)` as JS supports natively date comparisons

File: source/plugins/poopmap/index.mjs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    })
+
+    const hours = {}
+    for (let i = 0; i < filteredPoops.length; i++) {
+      const poop = filteredPoops[i]
+      const hour = new Date(poop.created_at).getHours()
+      if (!hours[hour]) hours[hour] = 1
```
Comment (lowlighter) [inline source/plugins/poopmap/index.mjs:27]:
```suggestion
      hours[hour] = (hours[hour] ?? 0) + 1
```
You can simplify these two lines with a single one

File: source/plugins/poopmap/index.mjs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      else hours[hour] += 1
+
+      if (!hours.max || hours[hour] > hours.max) hours.max = hours[hour]
```
Comment (lowlighter) [inline source/plugins/poopmap/index.mjs:30]: You should just initialize `const hours = {max:0}` and perform a `hours.max = Math.max(hours.max, hours[hour])` instead I think

File: source/templates/classic/partials/poopmap.ejs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      <% } else { %>
+        <section class="column chart largeable">
+          <h3>Poops per hour of day</h3>
+          <small>over the last <%- plugins.poopmap.days %> days</small>
+          <div class="chart-bars">
+            <% let displayedValues = []; %>
+            <% let poops = plugins.poopmap.poops %>
+            <% for (let h = 0; h < 24; h++) { displayedValues.push([h, (poops[h] || 0)/(poops.max || 1), poops[h]]) } %>
+  
+            <% for(let i = 0; i < displayedValues.length; i++) { %>
```
Comment (lowlighter) [inline source/templates/classic/partials/poopmap.ejs:20]:
```suggestion
            <% for(const [h, percentage, value] of displayedValues.length { %>
```
Is clearer than `for let` when you don't actually need index

**Review by lowlighter — CHANGES_REQUESTED** [review]
Hi thanks for your contribution 🙂 !

Looks good overall though there are some code parts that can be improved

You'll also need to provide a live-render of your plugin by running your branch on your own repository rather than a screenshot

File: source/templates/classic/partials/poopmap.ejs (hunk 5 above)
Comment (spenserblack) [reply source/templates/classic/partials/poopmap.ejs:20]: >`of displayedValues.length`

Did you mean `of displayedValues` instead?

Comment (lowlighter) [reply source/templates/classic/partials/poopmap.ejs:20]: yeah indeed 😅

File: source/plugins/poopmap/README.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  <img src="https://i.imgur.com/CctdShL.jpg" width="150" />
+</div>
+
+You should have something like
+```
+Haha, check out the places I've pooped on Poop Map https://api.poopmap.net/map?token=75f9b1f1af8372b8b991fb4e597db5f3
+```
+copied.
+
+Extract the `token` query paramater from the link. For example mine is: `75f9b1f1af8372b8b991fb4e597db5f3`
```
[the remark this answers is not in the document]
Comment (matievisthekat) [reply source/plugins/poopmap/README.md:41]: I understand, although these tokens are publicly shared (when sharing a profile for example)

File: source/plugins/poopmap/README.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
+### 💩 PoopMap plugin
+
+The *poopmap* plugin displays statistics from your [PoopMap](https://poopmap.net) account
+
+<table>
+  <td align="center">
+    <img src="">
```
[the remark this answers is not in the document]
Comment (matievisthekat) [reply source/plugins/poopmap/README.md:7]: A render of the plugin on its own? Is that just achieved but running the action with only the poopmap plugin enabled?

**Comment by matievisthekat** [comment]
The suggested edits have been resolved and the tests all pass now.

**Review by lowlighter — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**